File: Python/107.py
"""
@no 107
@name Binary Tree Level Order Traversal II
"""
class Solution:
    def getDepthOfTree(self, node):
        if node:
            return max(self.getDepthOfTree(node.left), self.getDepthOfTree(node.right)) + 1
        else:
            return 0

    def append(self, l, node, location):
        if node:
            l[location].append(node.val)
            self.append(l, node.left, location + 1)
            self.append(l, node.right, location + 1)

    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        l = []
        depth = self.getDepthOfTree(root)
        for i in range(0, depth):
            l.append([])

        location = 0
        self.append(l, root, location)
        return l[::-1]

File: JavaScript/107.js
/**
 * @no 107
 * @name Binary Tree Level Order Traversal II
 */
var levelOrderBottom = function(root) {
    let result = [];
    iteration(root, 1, result);
    return result.reverse();
};

var iteration = function(root, level, result) {
    if (!root) return;
    
    if (level > result.length) {
        result.push([]);
    }

    result[level - 1].push(root.val);
    iteration(root.left, level + 1, result);
    iteration(root.right, level + 1, result);
};